Conda's Prefixes Blogpost (#148)
conda-incubator/conda-dot-org: blog/2024-05-08-prefixes.mdx

revolved around installing packages into a singular environment, at the time
referred to as the `root` environment. This era, marked by `conda` versions
preceding 4.2.0, laid the groundwork for `conda`'s subsequent evolution as an
environment manager.

With the advent of [`conda 4.2.0`][conda-4.2.0] and beyond, a paradigm shift
occurred when the `conda-env` project was [merged][conda-env-pr] with the
`conda` project, elevating `conda` from a mere package manager to a
comprehensive environment manager. This pivotal transformation bestowed upon
`conda` the capability to create and manage separate environments, fostering
modularity and flexibility in development workflows.

Delving into `conda`'s Git history via `git blame` provides invaluable insights
into the timeline of the prefix system evolution. By tracing the commit history,
we unearth pivotal moments that introduced and refined the various prefixes
integral to `conda`'s architecture. The history is also revealing in how
non-linear the development of the prefix system has been, explaining many of the
complexities of trying to wrestle with the prefixes or trying to find ways to

> simplify them.

Reviewer: I'm having a hard time with this sentence. What I believe you're saying is that the non-linearity of the development of the prefix system explains how complex it has become and also explains some of the ways the devs have attempted to simplify it. It took me a few reads to get there. Is that correct?

Author: Fair and yes, that's what I was getting at. Let me take another stab at it.

1. `conda.base.context.context.conda_prefix`

   Introduced in `conda 4.2.0` (see https://github.com/conda/conda/commit/6e730cc)

>    as part of the private environment effort which has since been abandoned and

Reviewer: ```suggestion
   as part of the private environment effort, which has since been abandoned and
```

   deprecated.

   The `conda_prefix` lies at the heart of the prefix system, denoting the
   installation directory of `conda` itself. Typically, this aligns with the
   location of the `base` environment and is equivalent to Python's `sys.prefix`
   value.

   :::info

>    Use this anytime you need to interact with the `conda` installation itself.

Reviewer: ```suggestion
   Use this any time you need to interact with the `conda` installation itself.
```

   :::

2. `conda.base.context.context.root_prefix`

   > Previously `conda.config.root_dir` → `conda.config.ROOT_DIR`.

   Introduced in `conda 1.1.0` (see https://github.com/conda/conda/commit/d2b370a)
   and was originally intended for testing purposes only.

   The `root_prefix` serves as a [configurable value][condarc], most easily
   configurable in the user's `~/.condarc` file. In the absence of user-defined
   preferences, it defaults to the `conda_prefix`.

   :::warning

   Avoid using this. This is a contender for deprecation.

   :::

3. `conda.base.context.context.active_prefix`

   Introduced in `conda 4.4.0` (see https://github.com/conda/conda/commit/cbbe81f)
   as part of cleaning up the `conda info` output.

   Reflecting the currently active environment, the `active_prefix` is determined

>    by the `$CONDA_PREFIX` environment variable which is set by the `conda activate`

Reviewer: ```suggestion
   by the `$CONDA_PREFIX` environment variable, which is set by the `conda activate`
```

   and `conda deactivate` commands.

   :::warning

   Avoid using this. Treat this as an internal value.

   :::

4. `conda.base.context.context.default_prefix`

   > Previously `conda.config.default_prefix` → `conda.config.DEFAULT_ENV_PREFIX` →
   > `conda.config.default_environment` → `conda.anaconda.default_environment`.

   Introduced in `conda 1.1.0` (see https://github.com/conda/conda/commit/acd8144)
   as part of a larger configuration overhaul.

   The `default_prefix` effectively aggregates `root_prefix` and `active_prefix`,

>    embodying the default environment in the absence of CLI overrides.

Reviewer: I'm not totally sure what the last part of this sentence this means.

Author: When we specify `--name` or `--prefix` on the CLI we are effectively overriding the `default_prefix`, I agree the wording here is not clear, any suggestions would be great otherwise I'll wordsmith some more.

Author: Realized this was too verbose. Trimmed.

Reviewer: It's mainly the "embodying the default environment" bit that was tripping me up. 

So it aggregates (or combines together) `root_prefix` and `active_prefix`. `root_prefix` is configurable, but defaults to `conda_prefix`, which is conda's install directory, while `active_prefix` is based on what environment you have active.

Yeah, I guess I'm not sure what the default environment is in this context.

Author: Your right, that didn't describe anything concretely. Corrected.

   :::warning

   Avoid using this. Treat this as an internal value.

   :::

5. `conda.base.context.context.target_prefix`

   > Previously `conda.base.context.context.prefix` →
   > `conda.cli.common.get_prefix()` → `conda.cli.utils.get_prefix()`.

   Introduced in `conda 1.3.0` (see https://github.com/conda/conda/commit/9fd0f4b)
   to abstract out the `--name`/`--prefix` logic from the CLI implementations.

>    The `target_prefix` is the final prefix and first reflects the user input via

Reviewer: The final prefix in your list of prefixes? In terms of versioning, it seems like it was introduced third.

I also feel like I'm reading final like "ultimate" here.

Author: Ah so the prefixes are discussed in order of importance, from the most basic/bottom of the stack to the highest level/top of the stack. So "final" here just indicates that this is the top of the stack. Yes, this is the "ultimate" prefix but that also seems too grand?

Reviewer: Oh, definitely, don't use "ultimate" here. I was just saying that's what the connotation felt like to me. And I didn't realize we were discussing them from bottom of the stack to top! That's cool. Maybe I missed that in your explanation?

   the CLI options, `--name` or `--prefix`. In the absence of CLI overrides, it
   defaults to the `default_prefix`.

   :::info

   Use this when you need to interact with the current environment.

   :::

> In essence, `conda` prefix system serves as the cornerstone of its environment

Reviewer: ```suggestion
In essence, the `conda` prefix system serves as the cornerstone of its environment
```

revolved around installing packages into a singular environment, at the time
referred to as the `root` environment. This era, marked by `conda` versions
preceding 4.2.0, laid the groundwork for `conda`'s subsequent evolution as an
environment manager.

With the advent of [`conda 4.2.0`][conda-4.2.0] and beyond, a paradigm shift
occurred when the `conda-env` project was [merged][conda-env-pr] with the
`conda` project, elevating `conda` from a mere package manager to a
comprehensive environment manager. This pivotal transformation bestowed upon
`conda` the capability to create and manage separate environments, fostering
modularity and flexibility in development workflows.

Delving into `conda`'s Git history via `git blame` provides invaluable insights
into the timeline of the prefix system evolution. By tracing the commit history,
we unearth pivotal moments that introduced and refined the various prefixes
integral to `conda`'s architecture. The history reveals two distinct phases in
the evolution of `conda`'s prefix system. The first phase occurred early on in
`conda 1.x` and the second phase much later in `conda 4.x` with the merging of
`conda-env` with `conda`.

> The five prefixes defined within conda, ordered from the lowest level (system
> level and least configurable) to the highest level (user level and most
> configurable):

Reviewer 2: ```suggestion
The five prefixes defined within conda, ordered from the lowest level (system-level
and least configurable) to the highest level (user-level and most configurable):
```

4. `conda.base.context.context.default_prefix`

   > Previously:
   >
   > `default_prefix`
   >
   > ← `conda.config.default_prefix`
   >
   > ← `conda.config.DEFAULT_ENV_PREFIX`
   >
   > ← `conda.config.default_environment`
   >
   > ← `conda.anaconda.default_environment`

   Introduced in `conda 1.1.0` (see https://github.com/conda/conda/commit/acd8144)
   as part of a larger configuration overhaul.

   The `default_prefix` is the same as `active_prefix` when an environment is

>    active, when no environment is active the default environment is returned

Reviewer: ```suggestion
   active. When no environment is active the default environment is returned
```
You're also welcome to use a semi-colon.
